tests/test_client.py

```diff
@@ -383,3 +384,123 @@ async def test_client_no_pending_calls_warnings_with_max_concurrent_outgoing_cal
                 tg.start_soon(client.publish, topic)
 
         assert caplog.record_tuples == []
+
+
+async def test_client_not_reentrant() -> None:
+    client = Client(HOSTNAME)
+
+    with pytest.raises(MqttReentrantError):  # noqa: PT012
+        async with client:
+            async with client:
+                ...
+
+
+async def test_client_reusable() -> None:
+    client = Client(HOSTNAME)
+
+    async with client:
+        await client.publish("task/a", "task_a")
+
+    async with client:
+        await client.publish("task/b", "task_b")
+
+
+async def test_client_connect_disconnect() -> None:
+    client = Client(HOSTNAME)
+
+    await client.connect()
+    await client.publish("connect", "connect")
+    await client.disconnect()
+
+
+async def test_client_reusable_message() -> None:
+    custom_client = Client(HOSTNAME)
+    publish_client = Client(HOSTNAME)
+
+    async def task_a_customer() -> None:
+        async with custom_client:
+            async with custom_client.messages() as messages:
+                await custom_client.subscribe("task/a")
+                async for message in messages:
+                    assert message.payload == b"task_a"
+                    return
+
+    async def task_b_customer() -> None:
+        async with custom_client:
+            ...
+
+    async def task_a_publisher() -> None:
+        async with publish_client:
+            await publish_client.publish("task/a", "task_a")
+
+    async with anyio.create_task_group() as tg:
+        tg.start_soon(task_a_customer)
+        await anyio.sleep(1)
```

Comment:
It's an anti-pattern to `sleep` like that to ensure that `task_a_customer` actually started before we proceed. :)

Instead, use `tg.start`. See this great example from the anyio docs: https://anyio.readthedocs.io/en/3.x/tasks.html#starting-and-initializing-tasks

Let me know if you need help with it. 👍